code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -498,43 +591,290 @@ func (sc *StorageSmartContract) addGenerateChallengesStat(tp time.Time,
 	}
 }
 
-func (sc *StorageSmartContract) generateChallenges(t *transaction.Transaction,
-	b *block.Block, _ []byte, balances c_state.StateContextI) (err error) {
+func (sc *StorageSmartContract) getAllocationForChallenge(
+	t *transaction.Transaction,
+	allocID string,
+	balances c_state.StateContextI) (alloc *StorageAllocation, err error) {
 
-	var tp = time.Now()
-	defer sc.addGenerateChallengesStat(tp, &err)
-
-	var stats = &StorageStats{}
-	stats.Stats = &StorageAllocationStats{}
-	//var statsBytes util.Serializable
-
-	err = balances.GetTrieNode(stats.GetKey(sc.ID), stats)
+	alloc, err = sc.getAllocation(allocID, balances)
 	switch err {
 	case nil:
 	case util.ErrValueNotPresent:
-		return nil
+		Logger.Error("client state has invalid allocations",
+			zap.Any("selected_allocation", allocID))
+		return nil, common.NewErrorf("invalid_allocation",
+			"client state has invalid allocations")
 	default:
-		// unexpected MPT error
+		return nil, common.NewErrorf("adding_challenge_error",
+			"unexpected error getting allocation: %v", err)
+	}
+
+	if alloc.Expiration < t.CreationDate {
+		return nil, nil
+	}
+	if alloc.Stats == nil {
+		return nil, nil
+	}
+	if alloc.Stats.NumWrites > 0 {
+		return alloc, nil // found
+	}
+	return nil, nil
+}
+
+type challengeInput struct {
+	cr          *rand.Rand
+	t           *transaction.Transaction
+	challengeID string
+}
+
+type challengeOutput struct {
+	alloc            *StorageAllocation
+	storageChallenge *StorageChallenge
+	blobberChallenge *BlobberChallenge
+	allocChallenge   *AllocationChallenge
+	blobberAlloc     *BlobberAllocation
+	error            error
+}
+
+func (sc *StorageSmartContract) asyncGenerateChallenges(
+	blobberChallengeList partitions.RandPartition,
+	validators partitions.RandPartition,
+	r *rand.Rand,
+	creationDate common.Timestamp,
+	data <-chan challengeInput,
+	result chan<- challengeOutput,
+	balances c_state.StateContextI) {
+
+	defer close(result)
+
+	for d := range data {
```

Comment:
When only creating one challenge we want to create it synchronously. In which case, there will be only one entry in data.

I suggest putting the content of this for loop in a separate function that can be called directly from generateChallenge instead of `asyncGenerateChallenges` and also from for loop.